### ISSUE TITLE
Allow booting the ISO from a file on disk

### ISSUE THREAD

**Issue by probonopd** [furybsd/furybsd-livecd#169]
Allowing to the ISO from a file on disk would be a huge advantage, since one then could boot FurybBSD alongside other Live ISOs, e.g., using grub2.

This is what I am currently using in grub2:

```
menuentry "FuryBSD boot direct" {
  set isofile=/boot/iso/FuryBSD-12.1-XFCE-2020042001.iso
  loopback loop $isofile
  kfreebsd (loop)/boot/kernel/kernel
  kfreebsd_module (loop)/data/ramdisk.ufs.gz
  set kFreeBSD.vfs.root.mountfrom=ufs:/dev/md0
}
```

It results in the kernel being loaded, but then it cannot mount `/dev/rd0`.

The `loopback loop $isofile` allows GRUB to load files from there (e.g,. the kernel and the ramdisk).

The operating system will not see this loop mount, so the operating system must itself search for and loop-mount the ISO. Linux Live ISOs do this in the initramfs, e.g., Ubuntu is using casper for this task.

So I would guess that FreeBSD/FuryBSD don't have this capability yet, but it might be possible to be added:

https://github.com/ventoy/Ventoy can boot FuryBSD successfully. So I suspect it might have to do with a special kernel module they are apparently using:

https://github.com/ventoy/Ventoy/tree/master/Unix/ventoy_unix_src/FreeBSD/geom_ventoy_src/12.x/sys

And there is possibly more, such as a custom ramdisk.

Do you understand how they achieve booting FuryBSD from an on-disk ISO, and could this capability be brought into FuryBSD itself so that a third-party tool like Ventoy would not be needed to boot FuryBSD from e.g., grub2?

**Comment by pkgdemon** [comment]
PXE boot with syslinux has the same issue creating the virtual media device needed at /dev/iso9660/FuryBSD.  FreeBSD PXE boot has the same issue.  I found an article once which hinted at a regression in FreeBSD.  I'll see if I can dig that up later.  I know some other tricks with ISO assembly to make PXE work but the same trick breaks UEFI because of the staging size limitations with that.  That's the best quick answer I can give.  I'll have to circle back to this since it is is a complicated subject.

**Comment by pkgdemon** [comment]
Another thought I had considered for last few days due to a geom corrupt cosmetic issue that was pointed out to me with flash media (freebsd also has same issue) was to ship a raw disk image in addition to ISO as well.  We already build a raw disk image I just don't set the boot bits yet.

**Comment by probonopd** [comment]
Wouldn't we still have a similar issue, the raw image also needs to be found and loop-mounted?

**Comment by pkgdemon** [comment]
> Wouldn't we still have a similar issue, the raw image also needs to be found and loop-mounted?

I don't think so.  I know that raw method worked for PXE with another private project I used this code for but I would have to try it and see.

**Comment by probonopd** [comment]
For me personally, the ability to throw a FuryBSD ISO onto an existing disk alongside many existing Linux Live ISOs lowered the barrier to entry to the point that I actually tried it out. So it was a real gamechanger for me and got me started with FuryBSD and *BSD in general. I could imagine that it be a gamechanger for others, too, so I think we should make the ability to boot FuryBSD ISOs from disk a priority.

Use cases:
* I already have a drive with GRUB and a lot of other Live ISOs, and I would like to add FuryBSD
* I want to put multiple Live ISOs with different versions/builds/flavors of FuryBSD on the same USB stick to be able to easily decide at boot time which one to boot

https://www.ventoy.net/en/compatible.html describes:

> (...) Ventoy will create a virtual disk from the iso file and boot it. But the virtual disk is only on BIOS(Legacy or UEFI) level. In most case it works only in bootloader (...)

So this is what the `loop` command in `grub` can do. (Does the FreeBSD bootloader have a similar command?)

> Most of the morden OS's kernels will use their driver to access hardware after boot, so the virtual disk will not be visible to them. (...) But with ventoy they will not find such medium because there is no such physical media and they don't know that they were booted from a virual disk.

Ubuntu e.g., searches disks for a file with a certain name that can be passed on the boot command line with `iso-scan/filename=...`. But that is still not optimal, since that information needs to be provided manually in the grub configuration file. It would be even nicer if there was a mechanism by which the bootloader would pass this information to the system.

Ventoy does just that:

> In fact, ventoy [saves the information to runtime memory](https://www.ventoy.net/en/doc_compatible_format.html) before booting and provides a tool to get the information. In this way, the OS can know that they were booted from a virtual disk created by ventoy and can also know which iso file in which disk is the right image. Based on that, the OS can find the source medium successfully and continue to boot. 

Right now, Ventoy seems to use a custom initrd and kernel module for FuryBSD. I can imagine it is cumbersome work for Ventoy to keep this current and working for every FuryBSD Live ISO.

Hence, the following questions come to mind:

* Can we get the FreeBSD bootloader to loop-mount the ISO, and to populate the [Ventoy Information Format](https://www.ventoy.net/en/doc_compatible_format.html)? (Maybe using Forth?)
* Can we make the FuryBSD Live ISO  "Ventoy Compatible" by using that information to loop-mount the correct file if this information is available?
* Or should we do it in an entirely different way (e.g., the one that will be used for PXE)?
